tls/s2n_ktls.c

```diff
@@ -0,0 +1,201 @@
+        /* retrieve the receive fd */
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
+    } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        RESULT_ENSURE_REF(conn->send_io_context);
+        /* retrieve the send fd */
+        const struct s2n_socket_write_io_context *peer_socket_ctx = conn->send_io_context;
+        *fd = peer_socket_ctx->fd;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_GUARD(s2n_ktls_validate_socket_mode(conn, ktls_mode));
+
+#ifdef S2N_PLATFORM_SUPPORTS_KTLS
+    /* Enable 'tls' ULP for the socket. https://lwn.net/Articles/730207 */
+    int fd = 0;
+    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+
+    if (ignore_ktls_ulp_for_testing == false) {
+        int ret = setsockopt(fd, SOL_TCP, TCP_ULP, S2N_TLS_ULP_NAME, S2N_TLS_ULP_NAME_SIZE);
+        if (ret < 0) {
+            RESULT_BAIL(S2N_ERR_KTLS_ULP);
+        };
+    }
+
+    /* TODO configure keys */
+
+    return S2N_RESULT_OK;
+#else
+    RESULT_BAIL(S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+#endif
+}
+
+/*
+ * kTLS has been enabled on the socket. Errors are likely to be fatal and
+ * unrecoverable.
+ *
+ * TODO initiate connection close on error
+ */
+S2N_RESULT s2n_ktls_configure_connection(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+
+    /* TODO mark kTLS enabled on the connection */
+
+    return S2N_RESULT_OK;
+}
+
+/*
+ * Since kTLS is an optimization, it is possible to continue operation
+ * by using userspace TLS if kTLS is not supported. Upon successfully
+ * enabling kTLS, we set connection->ktls_send_enabled (and recv) to true.
+ *
+ * For this reason we categorize kTLS errors into recoverable and
+ * un-recoverable and handle them appropriately:
+ *
+ * - Errors related to the socket configuration are considered recoverable
+ *   since kTLS related `setsockopt` operations are non-destructive.
+ *
+ * - Errors related to connection configuration are considered
+ *   un-recoverable since we attempt to modify s2n_connection state.
+ */
+int s2n_ktls_enable(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_GUARD_RESULT(s2n_ktls_validate(conn));
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV || ktls_mode == S2N_KTLS_MODE_DUPLEX) {
+        POSIX_GUARD_RESULT(s2n_ktls_configure_socket(conn, S2N_KTLS_MODE_RECV));
+        POSIX_GUARD_RESULT(s2n_ktls_configure_connection(conn, S2N_KTLS_MODE_RECV));
+    }
+
+    if (ktls_mode == S2N_KTLS_MODE_SEND || ktls_mode == S2N_KTLS_MODE_DUPLEX) {
+        POSIX_GUARD_RESULT(s2n_ktls_configure_socket(conn, S2N_KTLS_MODE_SEND));
+        POSIX_GUARD_RESULT(s2n_ktls_configure_connection(conn, S2N_KTLS_MODE_SEND));
+    }
+
+    return S2N_SUCCESS;
+}
+
+/* Use for testing only */
```

Comment:
Might help to have a bit more documentation on why you would want to call this method in a test